### ISSUE TITLE
integrate with Dapr as Dapr's gateway

### ISSUE THREAD

**Issue by geffzhang** [microsoft/reverse-proxy#967]
Both the original and updated eShop application leverage the Envoy proxy as an API gateway, In the original eShopOnContainers implementation, the Envoy API gateway forwarded incoming HTTP requests directly to aggregator or back-end services. In the new eShopOnDapr, the Envoy proxy forwards the request to a Dapr sidecar.
I want to replace Enovy with Yarp  forwards the request to a Dapr sidecar 。

**Comment by jkotalik** [comment]
I'm not familiar with any envoy integrations with dapr, but are you effectively saying that you'd like to replace envoy with YARP as the reverse proxy? Does dapr have any unique integrations with envoy to make it work with Dapr?

I see these configurations [here](https://github.com/dotnet-architecture/eShopOnDapr/blob/1cd9106cf2475e33d6ce64ea60a2bc1a963e0250/src/ApiGateways/Envoy/envoy.yaml) and [here](https://github.com/dotnet-architecture/eShopOnDapr/blob/1cd9106cf2475e33d6ce64ea60a2bc1a963e0250/deploy/k8s/apigateway.yaml). The former being the configuration for envoy and the latter for k8s.

**Comment by geffzhang** [comment]
I have implemented a demo [TyeAndYarp](https://github.com/geffzhang/TyeAndYarp) and I expect yarp to have a better implementation that directly supports dapr ,this is main code:
https://github.com/geffzhang/TyeAndYarp/blob/master/ReverseProxy/DaprTransformProvider.cs

**Comment by bradygaster** [comment]
Adding @nishanil and @jamesmontemagno here. I think it would be exciting to see a YARP-inclusive version of eShop.

**Comment by karelz** [comment]
Triage: @samsp-msft can you please describe what is the plan here? (sample?)

**Comment by samsp-msft** [comment]
Plan is a proof of concept using YARP as a gateway into a set of DAPR services. based on the results of the POC we can decide if it should be a sample etc.

**Comment by ggleo08** [comment]
hi @samsp-msft , is there any progress about this feature ?

**Comment by samsp-msft** [comment]
Sorry- No progress as of yet.

**Comment by Arcalise08** [comment]
Almost a year later. Dapr has become even more dominant in the microservice space with Microsoft adopting it for Azure container apps. Easy integration (or really any type of integration) is essential for the health of a reverse proxy. 

We have got this working in production with Ocelot by using our Dapr service IDs as host names in combination with Daprs own delegating handler and it works perfectly fine. 

I'm just starting to look into it with YARP since Ocelot isnt being maintained anymore. But it looks like YARP doesnt directly expose anything that would make the integration easy like an `.AddDelegatingHandler`

**Comment by nikneem** [comment]
I solved this problem by addressing the (internal)container name, and not use DAPR. I really love the Azure Container Apps and use YARP as a reverse proxy to my underlying API's. This https://github.com/nikneem/tinylnk-gateway is my solution so far and to be honest, works like a charm. But still, you have to stay within a K8s context for this to work.

**Comment by Arcalise08** [comment]
> I solved this problem by addressing the (internal)container name, and not use DAPR. I really love the Azure Container Apps and use YARP as a reverse proxy to my underlying API's. This https://github.com/nikneem/tinylnk-gateway is my solution so far and to be honest, works like a charm. But still, you have to stay within a K8s context for this to work.

Yeah, hard coding service addresses most definitely not the way. My initial look at this seems to show it is relatively trivial to at least get a proof of concept working with yarp/dapr integration. This is what it looks like right now.

First setup a dapr delegating handler.

`public class DaprDelegatingHandler : DelegatingHandler
{
    private readonly ConcurrentDictionary<string, HttpClient> _httpClients = new();
    public DaprDelegatingHandler(HttpMessageHandler handler)
    {
        InnerHandler = handler;
    }

    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
        CancellationToken cancellationToken)
    {
        var daprEndpoint = request.RequestUri?.Host;
        if (string.IsNullOrEmpty(daprEndpoint) || !Utilities.IsValidDaprEndpoint(daprEndpoint))
            return base.SendAsync(request, cancellationToken);
        var client = _httpClients.GetOrAdd(daprEndpoint, key => DaprClient.CreateInvokeHttpClient(key));
        return client.SendAsync(request, cancellationToken);
    }
}`

Next create a ForwarderHttpFactory and use the delegating handler.

 
`public class DaprForwarderHttpClientFactory : IForwarderHttpClientFactory
{
    public HttpMessageInvoker CreateClient(ForwarderHttpClientContext context)
    {
        var handler = new SocketsHttpHandler
        {
            UseProxy = false,
            AllowAutoRedirect = false,
            AutomaticDecompression = DecompressionMethods.None,
            UseCookies = false,
            ActivityHeadersPropagator = new ReverseProxyPropagator(DistributedContextPropagator.Current)
        };
        return new HttpMessageInvoker(new DaprDelegatingHandler(handler));
    }
}`

Then in your service setup, just add the IForwarderHttpClientFactory

`services.AddSingleton<IForwarderHttpClientFactory, DaprForwarderHttpClientFactory>();`

It seems to work fine for simple service-to-service invocation calls. I cant promise it works in every situation. This works by setting your destinations as your dapr ID. So if your service dapr ID is "users" then your destination address should be "http://users"

**Comment by Arcalise08** [comment]
Heres the same thing as above with direct forwarding.

```csharp
services.AddHttpForwarder();
services.AddSingleton<IForwarderHttpClientFactory, DaprForwarderHttpClientFactory>();
var app = builder.Build();
app.Map("/api/users", async (HttpContext context, IHttpForwarder forwarder, IForwarderHttpClientFactory factory) =>
    await forwarder.SendAsync(context, "http://users", factory.CreateClient(new ForwarderHttpClientContext()))
);
```
If anyone has any feedback on this I'd be happy to hear it. It's using the same clientfactory as the above method uses. I'm unsure how to go about getting the `ForwarderHttpClientContext` so I have to create a new one, which seems to work fine considering my client factory doesn't use the context at all.

**Comment by MihaZupan** [comment]
> `factory.CreateClient(new ForwarderHttpClientContext())`

This will create a new HttpClient for every request.
If you're already using custom logic there, you can drop the `IForwarderHttpClientFactory` indirection and create your handler once upfront.
